File: packages/server-php/src/Session.php
<?php

declare(strict_types=1);

namespace OpenPasskey\Server;

use OpenPasskey\Base64Url;

class Session
{
    private const MIN_SECRET_LENGTH = 32;

    public static function validateConfig(SessionConfig $config): void
    {
        if (strlen($config->secret) < self::MIN_SECRET_LENGTH) {
            throw new \InvalidArgumentException(
                'session secret must be at least ' . self::MIN_SECRET_LENGTH . ' characters',
            );
        }
    }

    public static function createToken(string $userId, SessionConfig $config): string
    {
        $expiresAt = (int) (microtime(true) * 1000) + ($config->durationSeconds * 1000);
        $payload = "{$userId}:{$expiresAt}";
        $signature = hash_hmac('sha256', $payload, $config->secret, true);
        $signatureB64 = Base64Url::encode($signature);
        return "{$payload}:{$signatureB64}";
    }

    public static function validateToken(string $token, SessionConfig $config): SessionTokenData
    {
        $lastColon = strrpos($token, ':');
        if ($lastColon === false) {
            throw new \ValueError('invalid session token');
        }

        $signatureB64 = substr($token, $lastColon + 1);
        $rest = substr($token, 0, $lastColon);

        $secondLastColon = strrpos($rest, ':');
        if ($secondLastColon === false) {
            throw new \ValueError('invalid session token');
        }

        $userId = substr($rest, 0, $secondLastColon);
        $expiresAtStr = substr($rest, $secondLastColon + 1);

        if (!ctype_digit($expiresAtStr)) {
            throw new \ValueError('invalid session token');
        }

        $expiresAt = (int) $expiresAtStr;
        $payload = "{$userId}:{$expiresAtStr}";

        $expectedSig = hash_hmac('sha256', $payload, $config->secret, true);
        $providedSig = Base64Url::decode($signatureB64);

        if (!hash_equals($expectedSig, $providedSig)) {
            throw new \ValueError('invalid session token');
        }

        $nowMs = (int) (microtime(true) * 1000);
        $graceMs = $config->clockSkewGraceSeconds * 1000;
        if ($nowMs > $expiresAt + $graceMs) {
            throw new \ValueError('session expired');
        }

        return new SessionTokenData($userId, $expiresAt);
    }

    public static function buildSetCookieHeader(string $token, SessionConfig $config): string
    {
        $parts = [
            "{$config->cookieName}={$token}",
            'HttpOnly',
            "Path={$config->cookiePath}",
            "Max-Age={$config->durationSeconds}",
            "SameSite={$config->sameSite}",
        ];

        if ($config->secure) {
            $parts[] = 'Secure';
        }

        if ($config->domain !== null) {
            $parts[] = "Domain={$config->domain}";
        }

        return implode('; ', $parts);
    }

    public static function buildClearCookieHeader(SessionConfig $config): string
    {
        $parts = [
            "{$config->cookieName}=",
            'HttpOnly',
            "Path={$config->cookiePath}",
            'Max-Age=0',
            "SameSite={$config->sameSite}",
        ];

        if ($config->secure) {
            $parts[] = 'Secure';
        }

        if ($config->domain !== null) {
            $parts[] = "Domain={$config->domain}";
        }

        return implode('; ', $parts);
    }

    public static function parseCookieToken(?string $cookieHeader, SessionConfig $config): ?string
    {
        if ($cookieHeader === null || $cookieHeader === '') {
            return null;
        }

        foreach (explode(';', $cookieHeader) as $cookie) {
            $cookie = trim($cookie);
            $prefix = $config->cookieName . '=';
            if (str_starts_with($cookie, $prefix)) {
                $value = substr($cookie, strlen($prefix));
                return $value === '' ? null : $value;
            }
        }

        return null;
    }
}

File: packages/server-rails/lib/open_passkey/session.rb
# frozen_string_literal: true

require "openssl"

module OpenPasskey
  module Session
    MIN_SECRET_LENGTH = 32

    class << self
      def validate_config(config)
        if config.secret.length < MIN_SECRET_LENGTH
          raise ArgumentError, "session secret must be at least #{MIN_SECRET_LENGTH} characters"
        end
      end

      def create_token(user_id, config)
        expires_at = (Process.clock_gettime(Process::CLOCK_REALTIME) * 1000).to_i + (config.duration_seconds * 1000)
        payload = "#{user_id}:#{expires_at}"
        signature = OpenSSL::HMAC.digest("SHA256", config.secret, payload)
        signature_b64 = Base64Url.encode(signature)
        "#{payload}:#{signature_b64}"
      end

      def validate_token(token, config)
        last_colon = token.rindex(":")
        raise PasskeyError.new("invalid session token") unless last_colon

        signature_b64 = token[(last_colon + 1)..]
        rest = token[0...last_colon]

        second_last_colon = rest.rindex(":")
        raise PasskeyError.new("invalid session token") unless second_last_colon

        user_id = rest[0...second_last_colon]
        expires_at_str = rest[(second_last_colon + 1)..]

        unless expires_at_str.match?(/\A\d+\z/)
          raise PasskeyError.new("invalid session token")
        end

        expires_at = expires_at_str.to_i
        payload = "#{user_id}:#{expires_at_str}"

        expected_sig = OpenSSL::HMAC.digest("SHA256", config.secret, payload)
        provided_sig = Base64Url.decode(signature_b64)

        unless OpenSSL.fixed_length_secure_compare(expected_sig, provided_sig)
          raise PasskeyError.new("invalid session token")
        end

        now_ms = (Process.clock_gettime(Process::CLOCK_REALTIME) * 1000).to_i
        grace_ms = config.clock_skew_grace_seconds * 1000
        if now_ms > expires_at + grace_ms
          raise PasskeyError.new("session expired")
        end

        SessionTokenData.new(user_id, expires_at)
      end
    end
  end
end
